### PR TITLE
move lock inside if clause

### PR DESCRIPTION
**Pull request by bkamins**
As suggested by @nalimilan this should improve performance a bit and still be thread safe.

### CODE REVIEW

**Comment by bkamins** [comment]
I have benchmarked it and it actually is a relevant change when you iterate `GroupedDataFrame` by `GroupKey`.

File: src/groupeddataframe/groupeddataframe.jl [hunk 1]
```diff
@@ -50,19 +50,23 @@ end
 function Base.getproperty(gd::GroupedDataFrame, f::Symbol)
     if f in (:idx, :starts, :ends)
         # Group indices are computed lazily the first time they are accessed
-        Threads.lock(gd.lazy_lock)
         if getfield(gd, f) === nothing
```
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:53]:
```suggestion
        # Do not lock when field is already initialized
        if getfield(gd, f) === nothing
```

File: src/groupeddataframe/groupeddataframe.jl [hunk 2]
```diff
@@ -50,19 +50,23 @@ end
...
-            gd.idx, gd.starts, gd.ends = compute_indices(gd.groups, gd.ngroups)
+            Threads.lock(gd.lazy_lock)
+            if getfield(gd, f) === nothing
```
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:55]:
```suggestion
            if getfield(gd, f) === nothing # Do not lock when field is already initialized
```

File: src/groupeddataframe/groupeddataframe.jl [hunk 3]
```diff
@@ -50,19 +50,23 @@ end
...
+                gd.idx, gd.starts, gd.ends = compute_indices(gd.groups, gd.ngroups)
+            end
+            Threads.unlock(gd.lazy_lock)
         end
-        Threads.unlock(gd.lazy_lock)
         return getfield(gd, f)::Vector{Int}
     elseif f === :keymap
         # Keymap is computed lazily the first time it is accessed
-        Threads.lock(gd.lazy_lock)
         if getfield(gd, f) === nothing
```
Comment (nalimilan) [inline src/groupeddataframe/groupeddataframe.jl:63]:
```suggestion
        if getfield(gd, f) === nothing # Do not lock when field is already initialized
```

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**